lib/3scale/core/application_referrer_filter.rb

```diff
@@ -25,7 +25,7 @@ def self.delete(service_id, application_id, value)
       end
 
       def self.base_uri(service_id, application_id)
-        "#{default_uri}#{service_id}/applications/#{application_id}/referrer_filters"
+        "#{default_uri}#{service_id}/applications/#{CGI::escape(application_id.to_s)}/referrer_filters"
```

Comment:
I wonder whether we should use `ERB::Util.url_encode` or `CGI::escape`.

```
[1] pry(main)> str = "asdasd asdas"
[3] pry(main)> require 'cgi'
[6] pry(main)> require 'erb'

[4] pry(main)> CGI.escape str
=> "asdasd+asdas"

[7] pry(main)> ERB::Util.url_encode str
=> "asdasd%20asdas"
```

I'm not sure what is the processing on the other side. So this is just a question. Not an objection.

Reply:
Rails `Hash#to_query` seems to be using `CGI.escape` internally, so probably it is fine. But good to test it anyway.

Reply:
Yeah, I was looking for the best suited encoder when working on https://github.com/3scale/pisoni/pull/31, and I ended up with `CGI::escape`, but I don't remember the reason :sweat_smile: 
Probably one of the points was that it was already being used in other parts of the library (and worked successfully), so I decided not to innovate. Or maybe there was a different reason.

If you are really interested, I can dig into it (although I would prefer to avoid it :grimacing: )

Reply:
I think it has to be tested which one would work properly with a space character in the key. No need to dig further because this is the only difference that I was able to identify.

Reply:
Well, yes, I had tested that `CGI::escape` works as expected (in a manual integration test).
However, I decided to investigate this further to clarify any doubts.

This `pisoni` client is used for apisonator API, so, I checked what apisonator does with encoded spaces.

1. its base, sinatra, unescapes URL-encoded values in captures like this: https://github.com/sinatra/sinatra/blob/v2.2.4/lib/sinatra/base.rb#L1069

```
[5] pry(main)> URI::Parser.new.unescape("space%20man")
=> "space man"
[6] pry(main)> URI::Parser.new.unescape("space+man")
=> "space+man"
```
As we can see this results in different values of the `:app_id` [capture](https://github.com/3scale/apisonator/blob/master/app/api/internal/utilization.rb#L4) in the utilization URL (that this PR fixes).

2. When the application is fetched from Redis, the key for Redis passes through [encoding](https://github.com/3scale/apisonator/blob/8e7ff4f28088a0b78c173c90ad267524e532b76a/lib/3scale/backend/storage_key_helpers.rb#L4-L6) that replaces all spaces with `+`.

This is why both `%20` and `+` have the same effect.

By the way, regarding whether or not `+` should be decoded to spaces by sinatra, there was actually some discussion about it. It was implemented at some point, but then [reverted](https://github.com/sinatra/sinatra/commit/b4d7d4bb1088f5a6caf1bfbbba80717db9232fe7). And finally, `URI.decode` was also changed to `URI.unescape` [here](https://github.com/sinatra/sinatra/commit/4a02a757fde29ba21d241c838c8114f64a393fae).

Reply:
To confirm, this means we should use `ERB::Util.url_encode` to match sinatra behavior, correct?

Reply:
@akostadinov Mmm... no :upside_down_face: 

> To confirm, this means we should use `ERB::Util.url_encode` to match sinatra behavior, correct?

How did you come to this conclusion based on the info above? :sweat_smile: 

So, to summarize, for the app ID `space man` here are the transformations:

| encoder                      | pisoni URL     | sinatra capture | redis key   |
|----|----|----|----|
| using `ERB::Util.url_encode` | `/space%20man` | `space man`     | `space+man` |
| using `CGI::escape`          | `/space+man`   | `space+man`     | `space+man` |

That's why it works - the outcome (the key to search for the application in Redis) is exactly the same. If anything, I'd prefer to have it with `+` directly, to avoid two transformations.

By the way, it is a hypothetical case, because we don't allow spaces for application IDs, see [here](https://github.com/3scale/porta/blob/404885a5a00937c1329c219bf03e1559c11b31af/app/models/cinstance.rb#L124-L125).

Reply:
Got it why it works.

The point in using `ERB::Util.url_encode` is that we use same standard to encode on client and decode on server.

What goes to redis from server is server's business. If server is modified to use a different encoding to database, that shouldn't affect how client communicates with server. e.g. server might be changed to write Base64 in database or whatever else.

Also I understand about spaces not allowed. But it is only for sanity's sake that we use same encoding on client as we do on server. Actually wait, ideally we use `URI::Parser.new.escape` instead of `ERB::Util.url_encode`. Although better avoid creating a new parser object every time :)

Reply:
@akostadinov OK, you convinced me that spaces need to be encoded as `%20`. Well, I was convinced by reading some blog posts and comments, specifically this one https://stackoverflow.com/a/29948396/955810 (and its source).
I'll fix this for all places where special characters may appear in the URL path.

Reply:
So what is better `URI::Parser.new.escape` or `ERB::Util.url_encode`?

Reply:
I'm not sure the blog is entirely correct. It claims that `/` is alright in query but it is listed as reserved in both - [rfc1738](https://datatracker.ietf.org/doc/html/rfc1738#section-3.3) and [rfc2396](https://datatracker.ietf.org/doc/html/rfc2396#section-3.4)

The common characters to be escaped are also from [rfc2396](https://datatracker.ietf.org/doc/html/rfc2396#section-2.4.3)

Anyway, it appears that `URI::Parser` is https://ruby-doc.org/stdlib-3.1.0/libdoc/uri/rdoc/URI/RFC2396_Parser.html

But still it doesn't escape the `/` character. Which is strange, or I misinterpret the standard somehow...
```
[8] pry(main)> URI::RFC2396_Parser.new.escape "assad/ads da"
=> "assad/ads%20da"
```

Reply:
While `ERB::Util.url_encode` uses under the hood [CGI.escapeURIComponent(string)](https://rubyapi.org/3.2/o/cgi/escape#method-i-escapeURIComponent) which is supposed to follow [RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986#section-3.4) which in turn supersedes [rfc2396](https://datatracker.ietf.org/doc/html/rfc2396#section-3.4) and clarifies it.

```
[6] pry(main)> ERB::Util.url_encode "assad/ads da"
=> "assad%2Fads%20da"
```

But funnily from that same spec:
>  The characters slash ("/") and question mark ("?") may represent data
>  within the query component.  Beware that some older, erroneous
>  implementations may not handle such data correctly when it is used as
>  the base URI for relative references (Section 5.1), apparently
>  because they fail to distinguish query data from path data when
>  looking for hierarchical separators.  However, as query components
>  are often used to carry identifying information in the form of
>  "key=value" pairs and one frequently used value is a reference to
>  another URI, it is sometimes better for usability to avoid percent-
>  encoding those characters.

Basically the rfc2396 parser not encoding them while rfc3986 parser encoding them against it's own recommendation.

I think the main issue is that there are no separate methods to encode/decode query, path and whole URL.

So it is a little unclear which parts of the URLs the different methods are supposed to handle.

In this case either option appears to be alright.

Reply:
This is all very confusing indeed :zany_face:

Reply:
Ok, agreed to your conclusion. 

In fact [RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986#section-3.4) specifically allows `?` in query part as I posted above and suggests that in the past it was only erroneously parsed wrong. But apparently Ruby implementation of [RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986#section-3.4) accounts for erroneous parsing in ruby world :)

Thank you for looking into this!

Reply:
What we want is to percent-encode according to RFC3986, which seems to be the current standard. But it seems ruby doesn't really support it: https://bugs.ruby-lang.org/issues/18822

Probably `ERB::Util.url_encode` is the best option

Reply:
@jlledom yes, I'll do a mass replacement of all encoding, once I confirm that `ERB::Util.url_encode` is really the way to go :sweat_smile: right now having issues with some other special characters, e.g. `\`

Reply:
last comment of https://bugs.ruby-lang.org/issues/18822 suggests to use 

> But it's not useless since RFC 2396 was a Draft Standard and was obsoleted and updated by RFC 3986 which is an Internet Standard as CGI.escapeURIComponent and CGI.unescapeURIComponent implements RFC 398

And what I see is that `ERB::Util.url_encode` uses it under the hood. We may use that directly. The only "issue" I see with that is that `/` and `?` are % encoded against the recommendation but not violating the standard.

Reply:
> And what I see is that `ERB::Util.url_encode` uses it under the hood. We may use that directly. The only "issue" I see with that is that `/` and `?` are % encoded against the recommendation but not violating the standard.

What recommendation are you referring to?

I think for path segments it makes total sense to %-encode `/` and `?`, because they have a special meaning and otherwise modify the path (terminating the current path segment).

Reply:
Ah, this is path segment, not query. Ok.

Reply:
@akostadinov @jlledom fixed in https://github.com/3scale/pisoni/pull/33/commits/59751c8194a998aaf3c1985a357cd23a0b47fae6

Now, I am not sure about this path traversal thing, 
https://sinatrarb.com/protection/
https://github.com/sinatra/sinatra/blob/v2.2.4/rack-protection/lib/rack/protection/path_traversal.rb#L28-L29

This doesn't not allow us to use `/` and `\` in IDs/keys. It decodes them before routing takes place (and replaces `\` with `/`) so the requests are not routed properly and result in 404.

However, I am not quite sure what potential security implications may disabling it have, e.g.
```
use Rack::Protection, :except => :path_traversal
```
Ideas?

Reply:
Could you elaborate what the issue is? Encoding of the element should ensure no unescaped slashes are present so we will get the value with the slashes after rack URL processing.

Reply:
@akostadinov the problem is what path traversal rack protection does with the path, see the [link](https://github.com/sinatra/sinatra/blob/v2.2.4/rack-protection/lib/rack/protection/path_traversal.rb#L28-L29) I posted above.

so, the ID `one/two\three` (in Ruby `"one/two\\three"`) is encoded by pisoni correctly into `one%2Ftwo%5Cthree`, but the path traversal converts it to `one/two/thee`, which is obviously wrong, 

```
/internal/services/500/applications/one%2Ftwo%5Cthree/keys/one%2Ftwo%5Cthree
=>
/internal/services/500/applications/one/two/thee/keys/one/two/thee
```

Reply:
I think actually disabling path traversal is not going to cause a security issue, because we do not support static files, so I guess we can disable it.

However, I've been trying to disable this on the internal API, with no luck so far :upside_down_face:

Reply:
Are `\` and `/` valid characters for app ids? IMO the right thing to do is to disallow those characters in app ids and keep path traversal enabled. I don't feel comfortable disabling security checks.

But if we already have app ids with slashes in the DB there's not much we can do. Disabling path traversal should be safe if we are sure Apisonator will never serve FS files, but how are you sure it won't? is there some config I can check?

Reply:
> However, I've been trying to disable this on the internal API, with no luck so far 🙃

Is this test of any help?

https://github.com/sinatra/sinatra/blob/b626e2d82c23b4fde0b51782fd32ca27ccde1d1a/test/settings_test.rb#L614-L620

Reply:
> Are `\` and `/` valid characters for app ids? IMO the right thing to do is to disallow those characters in app ids and keep path traversal enabled. I don't feel comfortable disabling security checks.

I had previously disallowed `/` in app IDs and left a note about it [here](https://github.com/3scale/porta/pull/3639). I had checked the production DB and confirmed that there were very few occurrences, and all - for old or test apps.

I didn't realize `\` would be problematic too, so it's currently allowed, but we can check the DB as well, and potentially disallow it.

> But if we already have app ids with slashes in the DB there's not much we can do. Disabling path traversal should be safe if we are sure Apisonator will never serve FS files, but how are you sure it won't? is there some config I can check?

I am not sure to be honest, I am trying to understand how sinatra is configured :grimacing:

Reply:
In the meanwhile... https://github.com/3scale/apisonator/pull/366

Reply:
I'm ok to disable this. Seems to affect just this endpoint and we don't seem to serve files. Also relying on this protection to serve file would be very irresponsible.